src/aggregated/circulating_supply/circulating_supply_provider.rs

```diff
@@ -0,0 +1,193 @@
+};
+use crate::db_adapters::blocks;
+use crate::models;
+use crate::models::aggregated::circulating_supply::CirculatingSupply;
+
+const DAY: std::time::Duration = std::time::Duration::from_secs(86400);
+
+// Compute circulating supply on a daily basis, starting from 13 Oct 2020
+// (Transfers enabled moment on the Mainnet), and put it to the Indexer DB.
+// Circulating supply is calculated by the formula:
+// total_supply - sum(locked_tokens_on_each_lockup) - sum(locked_foundation_account)
+// The value is always computed for the last block in a day (UTC).
+pub(crate) async fn run_circulating_supply_computation(
+    view_client: Addr<near_client::ViewClientActor>,
+    pool: Database<PgConnection>,
+) {
+    let retry_duration = DAY.div(12);
+
+    // We perform actual computations 00:10 UTC each day to be sure that the data is finalized
+    let mut day_to_compute = lockup::TRANSFERS_ENABLED
+        .sub(Duration::from_secs(
+            lockup::TRANSFERS_ENABLED.as_secs() % DAY.as_secs(),
+        ))
+        .add(DAY)
+        .add(Duration::from_secs(10 * 60));
+
+    loop {
+        let now = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .expect("Time went backwards");
+
+        if now < day_to_compute {
+            time::sleep_until(Instant::now().add(day_to_compute.sub(now))).await;
+        }
+
+        match check_and_collect_daily_circulating_supply(&view_client, &pool, &day_to_compute).await
+        {
+            Ok(_) => {
+                day_to_compute = day_to_compute.add(DAY);
+            }
+            Err(err) => {
+                error!(
+                    target: crate::AGGREGATED,
+                    "Failed to compute circulating supply for {}: {}. Retry in {} hours",
+                    NaiveDateTime::from_timestamp(day_to_compute.as_secs() as i64, 0).date(),
+                    err,
+                    retry_duration.as_secs() / 60 / 60,
+                );
+                time::sleep(retry_duration).await;
+            }
+        };
+    }
+}
+
+async fn check_and_collect_daily_circulating_supply(
+    view_client: &Addr<near_client::ViewClientActor>,
+    pool: &Database<PgConnection>,
+    request_datetime: &Duration,
+) -> Result<Option<CirculatingSupply>, String> {
+    let start_of_day = request_datetime.as_nanos() - request_datetime.as_nanos() % DAY.as_nanos();
+    let latest_block_height = blocks::latest_block_height(&pool).await?.unwrap_or(0);
+
+    let block = blocks::get_latest_block_before_timestamp(&pool, start_of_day as u64).await?;
+    let block_timestamp = block
+        .block_timestamp
+        .to_u64()
+        .ok_or("`block_timestamp` expected to be u64")?;
+    let block_height = block
+        .block_height
+        .to_u64()
+        .ok_or("`block_height` expected to be u64")?;
+
+    if latest_block_height <= block_height {
+        return Err("Requested block is not finalized".to_string());
+    }
```

Comment:
This does not make a lot of sense since you query Postgres for both `latest_block_height` and `block_height` while the local node is what you want to check for the latest available block.

Query view-client for the final block to see what it is on the local node that indexer is running behind the scenes: https://github.com/FroVolod/near-cli/blob/8c37ede9136e13737d9ff0312f04f472877dd4dd/src/commands/view_command/view_account/block_id/mod.rs#L100-L108

P.S. I would probably move this check up to `run_circulating_supply_computation` since this has nothing to do with the circulating supply computation.